src/core/collector/mod.rs

```diff
@@ -0,0 +1,757 @@
+			let candidates = self.state.candidates_seen.get(para_id);
+			let disputes_concluded = self.state.disputes_seen.get(para_id).map(|disputes_seen| {
+				disputes_seen
+					.iter()
+					.filter(|dispute_info| dispute_info.concluded.is_some())
+					.cloned()
+					.collect::<Vec<_>>()
+			});
+
+			for channel in channels {
+				channel.send(CollectorUpdateEvent::NewHead(NewHeadEvent {
+					relay_parent_hashes: self.state.current_relay_chain_block_hashes.clone(),
+					relay_parent_number: self.state.current_relay_chain_block_number,
+					candidates_seen: candidates.cloned().unwrap_or_default(),
+					disputes_concluded: disputes_concluded.clone().unwrap_or_default(),
+					para_id: *para_id,
+				}))?;
+			}
+		}
+
+		for broadcast_channel in self.broadcast_channels.iter_mut() {
+			for (para_id, candidates) in self.state.candidates_seen.iter() {
+				let disputes_concluded = self.state.disputes_seen.get(para_id).map(|disputes_seen| {
+					disputes_seen
+						.iter()
+						.filter(|dispute_info| dispute_info.concluded.is_some())
+						.cloned()
+						.collect::<Vec<_>>()
+				});
+				broadcast_channel.send(CollectorUpdateEvent::NewHead(NewHeadEvent {
+					relay_parent_hashes: self.state.current_relay_chain_block_hashes.clone(),
+					relay_parent_number: self.state.current_relay_chain_block_number,
+					candidates_seen: candidates.clone(),
+					disputes_concluded: disputes_concluded.clone().unwrap_or_default(),
+					para_id: *para_id,
+				}))?;
+			}
+		}
+
+		self.state.candidates_seen.clear();
+		self.state.current_relay_chain_block_hashes.clear();
+		self.state.current_relay_chain_block_number = block_number;
+		self.state.current_relay_chain_block_hashes.push(block_hash);
+		Ok(())
+	}
+
+	/// Send event to all open channels
+	async fn broadcast_event(&mut self, event: CollectorUpdateEvent) -> color_eyre::Result<()> {
+		for (_, channels) in self.subscribe_channels.iter_mut() {
+			for channel in channels {
+				channel.send(event.clone())?;
+			}
+		}
+
+		for broadcast_channel in self.broadcast_channels.iter_mut() {
+			broadcast_channel.send(event.clone())?;
+		}
+
+		Ok(())
+	}
+
+	async fn process_new_head(&mut self, block_hash: H256) -> color_eyre::Result<()> {
+		let executor = self.api_service.subxt();
+		let ts = executor.get_block_timestamp(self.endpoint.clone(), Some(block_hash)).await;
+		let header = executor
+			.get_block_head(self.endpoint.clone(), Some(block_hash))
+			.await
+			.ok_or_else(|| eyre!("Missing block {}", block_hash))?;
+		let block_number = header.number;
+		info!(
+			"imported new block hash: {:?}, number: {}, previous number: {}, previous hashes: {:?}",
+			block_hash,
+			block_number,
+			self.state.current_relay_chain_block_number,
+			self.state.current_relay_chain_block_hashes
+		);
+
+		match block_number.cmp(&self.state.current_relay_chain_block_number) {
+			Ordering::Greater => {
+				self.update_state(block_number, block_hash)?;
```

Comment:
To avoid any possible racing vs consumers, update must to be sent only after we write everything related to this block in storage.

Reply:
Updates for the current block are sent when we receive a block with higher height. 

I don't know if we can receive events that belong to a block with lower height after we have seen a block with some current height that is more than that. It doesn't seem possible on a finalized chain at least.